www + d: Move comments likes tally from d to www

fernandoabolafio opened the pull request:
This PR removes the comment likes tally from d and add into www. It also standardizes the terminology as comment likes instead of mixing comment votes and comment likes across structs and functions.

The comment likes are all fetched from d when the www inventory is initialized and it is updated as new "comment likes" come in.
closes #603

lukebp commented on politeiawww/comments.go:
```diff
@@ -147,3 +160,61 @@ func validateComment(c www.NewComment) error {
+		comment.TotalVotes++
+	} else if lastUserVoteAction == newUserVoteAction {
+		// new action is equals the previous one, so we
+		// rever last user action
```
```suggestion
		// refer last user action
```

lukebp commented on politeiawww/backend.go:
```diff
@@ -90,6 +90,9 @@ type backend struct {
 	numOfPropsByUserID map[string]int
+
+	// User vote action on each comment
+	userVoteActionByCommentID map[string]map[string]int64 // [token+userid][commentid]action
```
We flip back and forth between referring to comment upvotes/downvotes as `votes` and `likes`.  For example, d has a `decredPluginCommentsLikesCache` and a www inventory record has a `commentLikes` field, but we refer them as `votes` right here.

Do you think it would make it easier to read if we were consistent and referred to them as `likes` everywhere instead of `votes`?

lukebp commented on politeiawww/inventory.go:
```diff
@@ -224,6 +225,45 @@ func (b *backend) loadRecord(record pd.Record) error {
+	if err != nil {
+		return fmt.Errorf("could not process comment likes results for %s: %v", t, err)
+	}
+	// b.calculateCommentsLikesScore(t)
```
```suggestion

```

lukebp commented on politeiawww/inventory.go:
```diff
@@ -224,6 +225,45 @@ func (b *backend) loadRecord(record pd.Record) error {
+
+	err = b.processCommentsLikesResults(t)
+	if err != nil {
+		return fmt.Errorf("could not process comment likes results for %s: %v", t, err)
```
80 column

lukebp commented on politeiawww/inventory.go:
```diff
@@ -379,6 +419,74 @@ func (b *backend) loadComments(t string) error {
+		return err
+	}
+
+	if err != nil {
```
remove

lukebp commented on politeiawww/inventory.go:
```diff
@@ -379,6 +419,74 @@ func (b *backend) loadComments(t string) error {
+		return err
+	}
+
+	payload, err := decredplugin.EncodeGetProposalCommentsLikes(decredplugin.GetProposalCommentsLikes{
```
80 column

lukebp commented on politeiawww/inventory.go:
```diff
@@ -379,6 +419,74 @@ func (b *backend) loadComments(t string) error {
+	}
+
+	// Decode plugin reply
+	gpclr, err := decredplugin.DecodeGetProposalCommentsLikesReply([]byte(reply.Payload))
```
80 column

lukebp commented on politeiawww/inventory.go:
```diff
@@ -379,6 +419,74 @@ func (b *backend) loadComments(t string) error {
+		b.inventory[token].commentsLikes[lc.CommentID] = append(cls, lc)
+	}
+
+	log.Tracef("loadCommentsLikes: %v inserted %v", token, len(gpclr.CommentsLikes))
```
80 column

lukebp commented on politeiawww/inventory.go:
```diff
@@ -224,6 +225,45 @@ func (b *backend) loadRecord(record pd.Record) error {
+
+			_, err := b._updateResultsForCommentLike(l)
+			if err != nil {
+				return fmt.Errorf("processCommentsLikesResults %v: updateResutsForCommentLike %v", token, err)
```
80 column

lukebp commented on politeiawww/inventory.go:
```diff
@@ -224,6 +225,45 @@ func (b *backend) loadRecord(record pd.Record) error {
+// resultant action per user per comment
+//
+// This function must be called WITH the mutext held
+func (b *backend) processCommentsLikesResults(token string) error {
```
If this needs to be called with the mutex held it should be named `_processCommentsLikesResults`, right?

lukebp commented on politeiawww/inventory.go:
```diff
@@ -379,6 +419,74 @@ func (b *backend) loadComments(t string) error {
+// of a given proposal
+//
+// This function must be called WITH the mutex held.
+func (b *backend) loadCommentsLikes(token string) error {
```
If this needs to be called with the mutex held it should be named `_loadCommentsLikes`, right?

fernandoabolafio commented on politeiawww/backend.go:
```diff
@@ -2108,10 +2110,11 @@ func (b *backend) ProcessLikeComment(lc www.LikeComment, user *database.User) (*
 		return nil, err
 	}
 
+	b.Lock()
```
I've done this because of the results of a comment like are now calculated on www. Before it was done on d, so the sequence was `validate like -> request d -> reply` now we have `validate like -> request d -> calculate like results -> reply`. 
Now that we have this calculation at the end of the function, I thought it would make sense to make the whole operation atomic to avoid the following situation:
- Two users voted up on a comment, user 1 see that the voting score is updated accordingly, user 2 see the voting score increased by two (his vote and the user 1 vote).

However, as per our discussion on slack, this race seems pretty benign as it doesn't compromise the data integrity and keeps the results up to date for the client. Therefore, I am reverting this and keeping the lock as it was before.

lukebp replied:
I agree.  The race is benign and preventing it is not worth the performance hit that we would take by locking during the politeiad call.

fernandoabolafio commented:
@lukebp I addressed all of your comments. I've also changed every reference from "comment vote" to "comment like" or the equivalent (e.g functions names, variables, etc) which required some changes on the GUI: https://github.com/decred/politeiagui/pull/911.

lukebp commented:
Code review looks good to me.  I tested this PR and it is working as expected.

marcopeereboom commented on politeiad/backend/gitbe/decred.go:
```diff
@@ -1408,11 +1353,7 @@ func (g *gitBackEnd) replayComments(token string) (map[string]decredplugin.Comme
-	if decredPluginCommentsUserCache != nil {
-		seen = decredPluginCommentsUserCache
-	}
+	commentsLikes := []decredplugin.LikeComment{}
```
Please use make here with the somewhat correct length.

fernandoabolafio replied:
I can make use of make here but It's not possible to know the correct length because the journals haven't been replayed at this point.

marcopeereboom replied:
Then use a PNOOMA.

fernandoabolafio replied:
From slack resolution:
```
moo31337 [3:12 PM]
because that requires a bunch of mallocs under the hood

if you have no idea about size pick something large like 1024
make(x,0,1024)
if you do know the size do: make(x,0,len(xxx))
```

marcopeereboom commented on politeiawww/backend.go:
```diff
@@ -2873,72 +2865,28 @@ func (b *backend) ProcessVoteStatus(token string) (*www.VoteStatusReply, error)
+	uclr := www.UserCommentsLikesReply{}
+	b.RLock()
+	defer b.RUnlock()
+	for k, v := range b.userLikeActionByCommentID {
```
This probably needs to go into a map. Do we have an idea how many iterations we do here?

fernandoabolafio replied:
It is a map, this is how it is defined:
```Go
// User vote action on each comment
userLikeActionByCommentID map[string]map[string]int64 // [token+userid][commentid]action
```
The number of iterations is equals to the number of comment votes a user has given to a specific proposal.

marcopeereboom replied:
I am talking about the for loop. What is a normal amount of iterations?

fernandoabolafio replied:
It depends on fo the number of users and proposals. The worse scenario is `NU x NP`,  being NU the total number of users and NP the total number of proposals.

fernandoabolafio replied:
From slack resolution:
```
fernandoabolafio [3:18 PM]
ahh, okay, that makes sense, thanks
2nd one:
You asked me how many iterations we could be doing for:
`for k, v := range b.userLikeActionByCommentID {...}`
So, the worst scenario here is a number of iterations equals to `totalNumberOfUsers x totalNumberOfProposals`. For the current live Pi I would estimate something about  1300  interactions (in the worst case) considering 13 proposals and 100 users.
Is that your question?

moo31337 [3:31 PM]
is this a one time deal?

fernandoabolafio [3:33 PM]
no, it happens as a result of an user request for their comment likes for a given proposal. E.g every time a logged in user access a proposal route.

moo31337 [3:34 PM]
so can we not iterate 1300 times?
that is an expensive lock

fernandoabolafio [3:35 PM]
Yeah, I can optimize that
```

marcopeereboom commented on politeiawww/comments.go:
```diff
@@ -147,3 +160,72 @@ func validateComment(c www.NewComment) error {
+// results and the vote resultant action for the user
+//
+// This function must be called WITH the mutex held
+func (b *backend) _updateResultsForCommentLike(like www.LikeComment) (*www.Comment, error) {
```
This looks expensive, can we optimize this a bit?

fernandoabolafio replied:
It is not an expensive call. Even though the code looks a bit complex, none of the operations run inside this function is costly. It grabs a value from the inventory, do some pretty basic math calculus and update that value in the inventory. So I think it is okay to make the entire function atomic.
What bits do you see as expensive for this function?

fernandoabolafio replied:
From slack resolution:
```
fernandoabolafio [3:35 PM]
Yeah, I can optimize that
okay, so the last one:
https://github.com/decred/politeia/pull/610/files#diff-c4d289a7e5b50d4535278b93a68b671aR178
You say that this function `_updateResultsForCommentLike(like www.LikeComment) (*www.Comment, error)` looks expensive.
I don’t think it is an expensive call. Even though the code looks a bit complex, none of the operations run inside this function is costly. It grabs a value from the inventory, do some pretty basic math calculus and update that value in the inventory. So I think it is okay to make the entire function atomic. What’s your concern? (edited)

moo31337 [3:53 PM]
ok
i can deal with that
```

marcopeereboom commented on politeiawww/inventory.go:
```diff
@@ -224,6 +225,48 @@ func (b *backend) loadRecord(record pd.Record) error {
+	commentLikes := ir.commentsLikes
+
+	for _, likes := range commentLikes {
+
```
No returns here please. Makes it hard to read.
